datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -466,10 +484,61 @@ public void onDatastreamUpdate() {
         }
       });
     }
+
+    if (_config.getEnableThroughputViolatingTopicsHandling()) {
+      _log.info(
+          "Populating the datastream violating topics to host level cache from the datastream objects on the update trigger");
+      try {
+        populateThroughputViolatingTopicsMap(datastreamGroups);
+      } catch (Exception exception) {
+        _log.error(
+            "received an exception while populating the datastream violating topics to host level cache from the "
+                + "datastream objects on the update trigger", exception);
+      }
+    }
     queueHandleAssignmentOrDatastreamChangeEvent(CoordinatorEvent.createHandleDatastreamChangeEvent(), true);
     _log.info("Coordinator::onDatastreamUpdate completed successfully");
   }
 
+  // This helper function populates the violations to local cache from the datastream metadata on every update call.
+  // Also note that updates to this local cache follows the behavior of replace-all, and not incremental.
+  private void populateThroughputViolatingTopicsMap(List<DatastreamGroup> datastreamGroups) {
+    _throughputViolatingTopicsMapWriteLock.lock();
+    try {
+      // clearing the cache as we would only maintain the latest reported information everytime.
+      _throughputViolatingTopicsMap.clear();
+
+      // fetching new violations from the datastream object.
+      datastreamGroups.forEach(datastreamGroup -> datastreamGroup.getDatastreams().forEach(datastream -> {
+        String commaSeparatedViolatingTopics = Objects.requireNonNull(datastream.getMetadata())
+            .get(DatastreamMetadataConstants.THROUGHPUT_VIOLATING_TOPICS);
+        if (Objects.nonNull(commaSeparatedViolatingTopics) && !commaSeparatedViolatingTopics.isEmpty()) {
+          String[] violatingTopics = commaSeparatedViolatingTopics.split(",");
+          if (violatingTopics.length > 0) {
+            _throughputViolatingTopicsMap.put(datastream.getName(), new HashSet<>(Arrays.asList(violatingTopics)));
+          }
+        }
+      }));
+    } finally {
+      _throughputViolatingTopicsMapWriteLock.unlock();
+    }
+  }
+
+  /**
+   * Fetch all the throughput violating topics per datastream from cache.
+   * */
+  public Set<String> getThroughputViolatingTopics(List<Datastream> datastreams) {
+    _throughputViolatingTopicsMapReadLock.lock();
```

Comment:
Does this need to be public? If you don't plan to expose this other than for tests, consider making it package private.

Reply:
done